barman/cloud_providers/google_cloud_storage.py

```diff
@@ -0,0 +1,245 @@
+
+    # Azure Blob Storage permit a maximum of 4.75TB per file
+    # This is a hard limit, while our upload procedure can go over the specified
+    # MAX_ARCHIVE_SIZE - so we set a maximum of 1TB per file
+    MAX_ARCHIVE_SIZE = 1 << 40
+
+    def __init__(self, url, jobs=2, encryption_scope=None):
+        """
+        Create a new Google cloud Storage interface given the supplied account url
+
+        :param str url: Full URL of the cloud destination/source (ex: )
+        :param int jobs: How many sub-processes to use for asynchronous
+          uploading, defaults to 2.
+        """
+        super(GoogleCloudInterface, self).__init__(
+            url=url,
+            jobs=jobs,
+        )
+        self.encryption_scope = encryption_scope
+
+        parsed_url = urlparse(url)
+        if not url.startswith(URL):
+            msg = f"google cloud storage URL {url} is malformed. Should start with '{URL}'"
+            raise ValueError(msg)
+        self.account_url = parsed_url.netloc
+        try:
+            self.bucket_name = parsed_url.path.split('/')[3]
+        except IndexError:
+            raise ValueError(f"Google cloud storage URL {url} is malformed. Bucket name not found")
+        path = parsed_url.path.split("/")[4:]
+        self.path = "/".join(path)
+
+        self.bucket_exists = None
+        self._reinit_session()
+
+    def _reinit_session(self):
+        """
+        Create a new session
+        Creates a client using "GOOGLE_APPLICATION_CREDENTIALS" env
+        """
+        # os.environ["GOOGLE_APPLICATION_CREDENTIALS"] = '/Users/didier.michel/Downloads/barman-324718-e759c283753d.json'
+        self.container_client = storage.Client().bucket(self.bucket_name)
+
+    def test_connectivity(self):
+        """
+        Test gcs connectivity by trying to access a container
+        """
+        try:
+            # We are not even interested in the existence of the bucket,
+            # we just want to see if google cloud storage is reachable.
+            self.bucket_exists = self._check_bucket_existence()
+
+            return True
+        except GoogleAPIError as exc:
+            logging.error("Can't connect to cloud provider: %s", exc)
+            return False
+
+    def _check_bucket_existence(self):
+        """
+        Check google bucket
+
+        :return: True if the container exists, False otherwise
+        :rtype: bool
+        """
+        return self.container_client.exists()
+
+    def _create_bucket(self):
+        """
+        Create the bucket in cloud storage
+        """
+        # Todo
+        #  There are several parameters to manage (
+        #  * project id ()
+        #  * data location (default US multi-region)
+        #  * data storage class (default standard)
+        #  * access control
+        #  Detailed documentation: https://googleapis.dev/python/storage/latest/client.html
+        # Might be relevant to use configuration file for those parameters.
```

Comment:
For AWS the region comes from the aws credentials file - hopefully the gcs credentials file allows something similar. For Azure the region is at the storage account level so has already been determined by the time we get to creating a container.

For everything else, the approach taken with both S3 and Azure was to choose a reasonable (i.e. secure) set of defaults and let people create their buckets themselves if they want anything more specific.

Reply:
Agreed, I'm using default parameters with gcp. If the user want something else, he can create it himself. I wrote some notes about this in the doc.